### PR TITLE
Update Symmetric Pointer Text

### PR DESCRIPTION
**Pull request by jdinan**
Clarify that OpenSHMEM routine arguments contain either symmetric or local addresses, where each is used, and restrictions on pointer arguments.

### CODE REVIEW

File: content/execution_model.tex [hunk 1]
```diff
@@ -52,3 +52,11 @@ \subsection{Progress of OpenSHMEM Operations}\label{subsec:progress}
...
       performance value for \openshmem programs.
 \end{itemize}
+
+\subsection{Invoking OpenSHMEM Operations}
+
+Pointer arguments to \openshmem routines that point to non-const data must not
+overlap in memory with other arguments to the same \openshmem operation.
+Otherwise, the behavior is undefined.  Two arguments overlap in memory if any
+of their data elements are located at the same address on the same \ac{PE}.
+
```
Comment (jdinan) [inline content/execution_model.tex]: Add: Some OpenSHMEM routines have multiple symmetric object arguments.  Unless noted, OpenSHMEM routines do not requires symmetric objects to be of the same kind, e.g. static versus dynamically allocated symmetric objects.

Comment (jdinan) [reply content/execution_model.tex]: @nspark Pointed out that the above doesn't adequately deal with shmem_ptr.

File: content/memory_model.tex [hunk 2]
```diff
@@ -52,6 +50,26 @@
...
+aligned based on their type and any requirements of the underlying
+architecture.  Symmetric addresses are only valid at the \ac{PE} where they
+were generated; using a symmetric address generated by a different \ac{PE}
+results in undefined behavior.
+
+The \FUNC{shmem\_ptr} routine allows the programmer to query a {\em local
+address} to a remotely accessible data object.  The resulting pointer is valid
+for direct load and store access; however, providing this address as an
+argument of an \openshmem routine that requires a symmetric address results in
+undefined behavior.
```
Comment (jdinan) [inline content/memory_model.tex]: @nspark Mention that every symmetric address is also a valid local address (i.e. local address is a superset).

File: content/memory_model.tex [hunk 3]
```diff
@@ -37,11 +35,41 @@
...
+Symmetric data objects are referenced in \openshmem operations through the
+local pointer to the desired remotely accessible object.  The address contained
+in this pointer is referred to as a {\em symmetric address}.  Every symmetric
+address is also a {\em local address} that is valid for direct memory access;
+however, not all local addresses are symmetric.  Manipulation of symmetric
+addresses passed to \openshmem routines, including pointer arithmetic;
+structure or union member access; and array indexing are permitted as long as
+the resulting local pointer remains within the same symmetric allocation or
+object.  Symmetric addresses are only valid at the \ac{PE} where they were
+generated; using a symmetric address generated by a different \ac{PE} results
```
Comment (agrippa) [inline content/memory_model.tex]: just a minor suggestion for clarity:

using a symmetric address generated by a different \ac{PE} for direct memory access or as an argument to an \openshmem routine results in undefined behavior.

File: content/memory_model.tex [hunk 4]
```diff
@@ -37,11 +35,41 @@
...
+in undefined behavior.
+
+Symmetric addresses provided to typed and type-generic \openshmem interfaces
+must be naturally aligned based on their type and any requirements of the
+underlying architecture.  Symmetric addresses provided to fixed-size (e.g.
+\FUNC{shmem\_put32}) \openshmem interfaces must also be aligned to the given
+size.  The ``mem'' interfaces, e.g. \FUNC{shmem\_putmem} have no alignment
+requirements.  Because \CorCpp{} structures may be contain
```
Comment (davidozog) [inline content/memory_model.tex]:
```suggestion
requirements.  Because \CorCpp{} structures may contain
```

File: content/shmem_atomic_add.tex [hunk 5]
```diff
@@ -35,11 +35,10 @@
...
         When this argument is not provided, the operation is performed on
         the default context.}
-    \apiargument{OUT}{dest}{The remotely accessible integer data object to be
-        updated  on the remote \ac{PE}.  When using \CorCpp, the type of
-        \dest{} should match that implied in the SYNOPSIS section.}
-    \apiargument{IN}{value}{The value to be atomically added to \dest. When using \CorCpp, the type of \VAR{value} should match that  implied  in
-        the SYNOPSIS  section.}
+    \apiargument{OUT}{dest}{Symmetric address of the destination data object.
+        The type of \dest{} should match that implied in the SYNOPSIS section.}
+    \apiargument{IN}{value}{The value to be atomically added to \dest.
```
Comment (davidozog) [inline content/shmem_atomic_add.tex]: `dest` is an address, so I'm not sure we're not "adding" `value` to that address.
Perhaps:
```suggestion
    \apiargument{IN}{value}{The value to be atomically added to the destination object.
```

File: content/shmem_atomic_fetch_add_nbi.tex [hunk 6]
```diff
@@ -24,10 +24,12 @@
...
       is performed on the default context.}
-  \apiargument{OUT}{fetch}{Local data object to be updated.}
-  \apiargument{OUT}{dest}{The remotely accessible data object to be updated on
-      the remote \ac{PE}.}
-  \apiargument{IN}{value}{The value to be atomically added to \VAR{dest}.}
+  \apiargument{OUT}{fetch}{Local address of data object to be updated.
+      The type of \VAR{fetch} should match that implied in the SYNOPSIS section.}
+  \apiargument{OUT}{dest}{Symmetric address of the destination data object.
+    The type of \dest{} should match that implied in the SYNOPSIS section.}
+  \apiargument{IN}{value}{The value to be atomically added to \VAR{dest}.
```
Comment (davidozog) [inline content/shmem_atomic_fetch_add_nbi.tex]: Adding `value` to destination object?

File: content/shmem_atomic_fetch_add.tex [hunk 7]
```diff
@@ -36,8 +36,8 @@
 \apiargument{IN}{ctx}{A context handle specifying the context on which to perform the operation.
     When this argument is not provided, the operation is performed on
     the default context.}
-\apiargument{OUT}{dest}{The remotely accessible integer data object to be updated on
-    the remote \ac{PE}. The type of \VAR{dest} should match that implied in the
+\apiargument{OUT}{dest}{Symmetric address of the destination data object.
+    The type of \VAR{dest} should match that implied in the
     SYNOPSIS section.}
 \apiargument{IN}{value}{The value to be atomically added to \VAR{dest}.  The
```
Comment (davidozog) [inline content/shmem_atomic_fetch_add.tex]: Perhaps:
```suggestion
\apiargument{IN}{value}{The value to be atomically added to the destination object.  The
```

File: content/execution_model.tex [hunk 8]
```diff
@@ -1,27 +1,33 @@
-An \openshmem program consists of a set of \openshmem processes called \acp{PE}
-that execute in an \ac{SPMD}-like model where each \ac{PE} can take a different
-execution path. For example, a \ac{PE} can be implemented using an OS
-process. The \acp{PE} may be either single or multithreaded.
-The \acp{PE} progress asynchronously, and can communicate/synchronize
-via the \openshmem interfaces.  All \acp{PE} in an \openshmem program should
-start by calling the initialization routine \FUNC{shmem\_init}%
+An \openshmem program consists of a set of \openshmem processes called
+\acp{PE}.  While not required by \openshmem, in typical usage \acp{PE} are
```
Comment (wrrobin) [inline content/execution_model.tex]: usage`,`

File: content/execution_model.tex [hunk 9]
```diff
@@ -52,3 +58,21 @@ \subsection{Progress of OpenSHMEM Operations}\label{subsec:progress}
       provide asynchronous one-sided operations, as these have very limited
       performance value for \openshmem programs.
 \end{itemize}
+
+\subsection{Invoking OpenSHMEM Operations}
+
+Pointer arguments to \openshmem routines that point to non-const data must not
+overlap in memory with other arguments to the same \openshmem operation, with
+the exception of in-place reductions as described in Section~\ref{subsec:shmem_reductions}.
```
Comment (nspark) [inline content/execution_model.tex]: Breadcrumb: allow in-place broadcast operations per #309

File: content/shmem_addr_accessible.tex [hunk 10]
```diff
@@ -10,23 +10,25 @@
...
-    specified address \VAR{addr} is a data object on the local \ac{PE}.
+    specified address \VAR{addr} is the local address of the data object on the
+    local \ac{PE}.
 }
 
 \apireturnvalues{
-    The return value is \CONST{1} if \VAR{addr} is a symmetric data object
-    and accessible via \openshmem routines from the specified remote \ac{PE};
-    otherwise, it is \CONST{0}.
+    The return value is \CONST{1} if \VAR{addr} is also a symmetric
```
Comment (wrrobin) [inline content/shmem_addr_accessible.tex]: Should we write instead: `if local address \VAR{addr} is also a symmetric address`? In its current form, `also` feels redundant.

File: content/execution_model.tex [hunk 11]
```diff
@@ -1,27 +1,33 @@
-An \openshmem program consists of a set of \openshmem processes called \acp{PE}
-that execute in an \ac{SPMD}-like model where each \ac{PE} can take a different
-execution path. For example, a \ac{PE} can be implemented using an OS
-process. The \acp{PE} may be either single or multithreaded.
-The \acp{PE} progress asynchronously, and can communicate/synchronize
-via the \openshmem interfaces.  All \acp{PE} in an \openshmem program should
-start by calling the initialization routine \FUNC{shmem\_init}%
+An \openshmem program consists of a set of \openshmem processes called
+\acp{PE}.  While not required by \openshmem, in typical usage \acp{PE} are
+executed using a single program, multiple data (\ac{SPMD}) model.  SPMD
```
Comment (manjugv) [inline content/execution_model.tex]: SPMD -> \ac{SPMD}

File: content/shmem_ptr.tex [hunk 12]
```diff
@@ -9,16 +9,21 @@
...
 
 \apidescription{
     \FUNC{shmem\_ptr} returns an address that may be used to directly reference
-    \dest{} on the specified \ac{PE}.  This address can be assigned to a pointer.
-    After that, ordinary loads and stores to this remote address may be performed.
-
+    \dest{} on the specified \ac{PE}.  This address can be assigned to a
+    pointer.  After that, ordinary loads and stores to \dest{} may be
+    performed.  The address returned by \FUNC{shmem\_ptr} is a local address to
+    a remotely accessible data object.  Providing this address to argument of
```
Comment (jdinan) [inline content/shmem_ptr.tex]: to an argument

File: content/memory_model.tex [hunk 13]
```diff
@@ -37,11 +35,41 @@
...
+structure or union member access; and array indexing are permitted as long as
+the resulting local pointer remains within the same symmetric allocation or
+object.  Symmetric addresses are only valid at the \ac{PE} where they were
+generated; using a symmetric address generated by a different \ac{PE} results
+in undefined behavior.
+
+Symmetric addresses provided to typed and type-generic \openshmem interfaces
+must be naturally aligned based on their type and any requirements of the
+underlying architecture.  Symmetric addresses provided to fixed-size (e.g.
+\FUNC{shmem\_put32}) \openshmem interfaces must also be aligned to the given
```
Comment (jdinan) [inline content/memory_model.tex]:
```suggestion
\FUNC{shmem\_put32}) \openshmem interfaces must have storage size equal to the bit-width of the given operation\footnote{The size of a byte is implementation-defined in C.  Users can consult \CONST{CHAR\_BIT} in limits.h for additional information} and be aligned to the given
```

Comment (jdinan) [reply content/memory_model.tex]: @nspark Feel free to help with wordsmithing.  :neckbeard:

File: content/shmem_atomic_add.tex (hunk 5 above)
Comment (jdinan) [reply content/shmem_atomic_add.tex]: Nice catch! I cribbed text from the new bitwise AMOs -- "The operand to the atomic add operation." Let me know if you prefer something different.

**Comment by jdinan** [comment]
Special ballot changes: 0cf97e5

File: content/shmem_atomic_add.tex (hunk 5 above)
Comment (davidozog) [reply content/shmem_atomic_add.tex]: Sounds pretty good.  If I were to nitpick, I might say that an add operation has 2 operands, so which one is this one?  :wink:  It's good though, I think it's pretty clear this is the sender's operand.

**Pull request merged**